### ISSUE TITLE
Basic Checker for Undefined Refs and Duplicate Field Names

### ISSUE THREAD

**Issue by amar47shah** [frontrowed/stratosphere#33]
from the project README:

>Future Work:
Implement basic checker for things like undefined Refs and duplicate field names. This stuff would be too unwieldy to do in types, and performing a checking pass over a template should be pretty straightforward.

I'd like to try my hand at this. Could y'all provide some details about:

1. the interface of the checker
2. specific cases, with input and output

Thanks!

**Comment by jdreaver** [comment]
Hey @amar47shah!

I think the checker would have a type of something like: `Template -> [CheckError]`, and `CheckerError` 
can be a sum type of various types of errors:

```haskell
data CheckerError
  = UndefinedRef Text
  | DuplicateFieldName Text
  ...
```

If there are no errors then of course we could return the empty list. It's up to the user to run the checker and to fail and/or print errors.

To find undefined `Ref`s, we have to gather all the allowed references from the `Parameters` section and from the names of all the `Resource`s. Then, we would somehow have to traverse the templates and get any uses of the `Ref` function, and then make sure that every `Ref` is inside our list of possible references. To be honest traversing the template and just extracting `Ref` names sounds really nontrivial without a ton of boilerplate and/or metaprogramming, and I think there might be a better strategy.

Finding duplicate resource names is pretty easy, just get the `resourceResName` for all the resources and see if there are duplicates. Right now if there are duplicates I think we silently discard the earlier occurrences in the process of JSON serialization.

Basically, do whatever you think is best! This isn't a core feature of `stratosphere`, so we can mark it as experimental or something without affecting the quality of users' lives. I do want to think carefully about the design though and make sure it is indeed useful lest we throw something half-baked into the codebase and confuse new users who wonder what it's for.

**Comment by amar47shah** [comment]
Thanks, @jdreaver! I'll dive in and let you know soon how it's going.

**Comment by amar47shah** [comment]
Took a crack at the easier problem of finding duplicates.

@jdreaver, let me know if this is on the right track:
https://github.com/amar47shah/stratosphere/blob/error-checker/library/Stratosphere/Check.hs

thanks!

**Comment by jdreaver** [comment]
Oh wow, that's a very point-free function! Haha.

Yeah that looks great. Even just checking for duplicate field names is useful. I'm wondering what kinds of checks we can do besides checking for undefined refs, because that is probably not possible without a ton of work (and subsequently not really worth it because CloudFormation tells you instantly when a Ref is undefined).

**Comment by amar47shah** [comment]
Thanks for having a look!

> checking for undefined refs, because that is probably not possible without a ton of work (and subsequently not really worth it

Would [this](http://stackoverflow.com/questions/27815489/is-it-possible-to-list-the-names-and-types-of-fields-in-a-record-data-type-that) be a worthwhile direction to pursue?

> I'm wondering what kinds of checks we can do

Are there any recurring problems that have come up in practice?

**Comment by jdreaver** [comment]
> Would this be a worthwhile direction to pursue?

Yeah, we'll have to do something similar to that, using Generics or Template Haskell to enumerate Ref values. We are also doing code generation from scratch, so we could find some class to implement that returns what we want. However, I don't want to do anything that will increase compilation time too much. Undefined Refs aren't too big of a problem in practice.

> Are there any recurring problems that have come up in practice?

To be honest, not many. Whenever I have an error it's usually because I gave something an improper value. Things like this can actually be fixed by having stricter types, not with a checker like this. I'll be sure to write something down here if I think of it though. If anyone else sees this issue I'm sure they can chime in.

**Comment by amar47shah** [comment]
> improper values can actually be fixed by having stricter types, not with a checker like this.

Makes sense. I'd be interested to hear more about which types in this project could benefit from being stricter.

Is this duplicate checker useful? Would it be worth it to open a PR now?

**Comment by jdreaver** [comment]
> I'd be interested to hear more about which types in this project could benefit from being stricter.

I usually try to fix them as I encounter them, so I don't have any off the top of my head :)

> Is this duplicate checker useful? Would it be worth it to open a PR now?

I think so. How about this: make a decent-sized comment explaining the purpose of the checker (catch errors that would be too expensive or unwieldy to encode in types), and saying it is experimental. That way new users won't get confused by what it is, and they won't think they have to use it.

Thanks for taking a crack at this!

**Comment by amar47shah** [comment]
PR opened. Thanks!

**Comment by jdreaver** [comment]
Oh, I just thought of another thing we can check: properties whos `Required` value is conditional based on other properties. For each resource type with conditionally required properties, we can build a function that checks that these conditional requirements are valid. (I'm just writing this down here for the future, no need to jump on this :smile:)

**Comment by jdreaver** [comment]
Closed via #36
